### ISSUE TITLE
Cannot construct recursive alues

### ISSUE THREAD

**Issue by rvs314** [coalton-lang/coalton#619]
In many statically typed languages, values can be constructed recursively, for example in the following ocaml code:

```ocaml
let rec k = 0 :: k
```

The equivalent coalton code doesn't work:

```lisp
(coalton 
  (let ((a (cons 0 a)))
    a))
```

Giving the following error:

```
Variable A-405 cannot be defined recursively.
In (LET ((A (CONS 0 A)))
  A)
In COALTON
   [Condition of type COALTON-IMPL/TYPECHECKER::SELF-RECURSIVE-VARIABLE-DEFINITION]
```

Mutually recursive definitions give a different error:

```lisp
(coalton
  (let ((a (cons 0 b))
        (b (cons 0 a)))
    a))
```

```
Invalid scc binding group. This should have been detected during typechecking.
   [Condition of type COMMON-LISP:SIMPLE-ERROR]
```

**Comment by karlosz** [comment]
Maybe adding `letrec` as a primitive is in order? Unlike scheme, it would have to imply lazy semantics (which I assume OCaml is doing).

**Comment by rvs314** [comment]
> Unlike scheme, it would have to imply lazy semantics (which I assume OCaml is doing).

I don't think this is the case; the following OCaml code fails:

```ocaml
let cons_zero v = 0 :: v
let rec k = cons_zero k
```

I believe they just have a special case for constructors

**Comment by gefjon** [comment]
> it would have to imply lazy semantics (which I assume OCaml is doing).

I don't believe OCaml has laziness for this case; I think they just do what I tried in PR [620]. That strategy is:

- Only allow `letrec` of constructors.
- Expand `letrec x = Ctor a b in body` into `let x = Ctor uninit uninit in (set! (Ctor-a x) a); (set! (Ctor-b x) b); body`

The only thing holding us back from doing that is that, in release mode, Coalton will optimize single-variant single-slot `define-types` to not allocate; i.e. for `(define-type Foo (Foo Bar))`, the constructor `Foo` is the identity function, and the representation of the type `Foo` is the same as `Bar`.
